### PR TITLE
fix(socket sink): Properly detect shutdown

### LINKED ISSUE #2406: Docker+Datadog very unreliable
This happens when running the vector-0.9.0-1.x86_64 RPM on 10 separate Centos 7 boxes in AWS. I'm using a docker source with a datadog sink. My vector.toml contains:

```
[sources.monitor_source]
  type               = "docker"
  include_containers = ["monitor-"]

[transforms.monitor_transform]
  type            = "add_fields"
  inputs          = ["monitor_source"]
  fields.hostname = "my-hostname"
  fields.service  = "monitor"
  fields.ddsource = "vector"
  fields.ddtags   = "monitor,env:production,project:my-project,location:london"

[sinks.monitor_sink]
  type           = "datadog_logs"
  inputs         = ["monitor_transform"]
  api_key        = MY_DATADOG_API_KEY
  encoding.codec = "json"
      
  [sinks.monitor_sink.tls]
    enabled = true
    ca_path = "/etc/ssl/certs/ca-bundle.crt"
```

This works briefly after I start vector running. I can see logs going into datadog, but it just stops working after a couple of logs have got in. I tried switching to using the syslog source and using the docker syslog driver but this didn't help. I tried adding a disk buffer to the sink and that didn't help. I tried without tls and that didn't help.

I know both of these components are in beta atm, so I'm just wondering if this is expected behaviour at this time, or if I'm seeing something different to what everyone else is?

### PR DESCRIPTION
**Pull request by LucioFranco**
This change fixes how we handle when the server side of the TCP
connection within the socket sink handles _proper_ shutdowns. A proper
shutdown is when the server side calls `shutdown(2)` with the `SHUT_WR`
flag. This then produces an EOF on the read side which we can detect to
start the reconnect process before we attempt to write the next event.

Currently, there is the possibility that an event will get dropped if we
don't detect this EOF and attempt to write. This has been fixed and we
now have a test to detect that we don't lose events if the server
shutsdown properly. There is still a case where if the read returns
`WouldBlock` and the write fails we will still lose events but this is
an improper shutdown and should be extremely rare.

Fixes #2406

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

File: src/internal_events/tcp.rs [hunk 1]
```diff
@@ -56,6 +56,23 @@ impl InternalEvent for TcpConnectionDisconnected {
...
+#[derive(Debug)]
+pub struct TcpConnectionShutdown {}
+
+impl InternalEvent for TcpConnectionShutdown {
+    fn emit_logs(&self) {
+        debug!(message = "received EOF from the server; reconnecting.");
+    }
+
+    fn emit_metrics(&self) {
+        counter!("tcp_connectio_shutdown", 1,
```
Comment (lukesteensen) [inline src/internal_events/tcp.rs:68]: `connectio`

File: src/sinks/socket.rs [hunk 2]
```diff
@@ -155,4 +155,106 @@ mod test {
...
+        let (sink, _) = rt.block_on(pump).unwrap();
+
+        // Loop and check for 10 events, we should always get 10 events. Once,
+        // we have 10 events we can tell the server to shutdown to simulate the
+        // remote shutting down on an idle connection.
+        for _ in 0..100 {
+            let amnt = counter.load(Ordering::SeqCst);
+
+            if amnt == 10 {
+                close.notify();
```
Comment (lukesteensen) [inline src/sinks/socket.rs:238]: Couldn't we just have the server count to ten and shutdown itself?

File: src/sinks/socket.rs [hunk 3]
```diff
@@ -155,4 +155,106 @@ mod test {
...
+                break;
+            }
+
+            std::thread::sleep(std::time::Duration::from_millis(100));
```
Comment (lukesteensen) [inline src/sinks/socket.rs:242]: I like to think we could do this without sleeping. Maybe very shortly at the end before checking the final count?

File: src/sinks/util/tcp.rs [hunk 4]
```diff
@@ -175,14 +175,33 @@ impl Sink for TcpSink {
...
+                // This type internally is either a `TcpStream` or a
+                // `SslStream<TcpStream>`. This match will provide the inner
+                // `TcpStream`.
+                let conn = match connection.get_mut() {
+                    MaybeTls::Raw(t) => t,
+                    MaybeTls::Tls(t) => t.get_mut().get_mut(),
+                };
+
+                // Test if the remote has issued a disconnect by calling read(2)
+                // with a 0 sized buffer.
```
Comment (lukesteensen) [inline src/sinks/util/tcp.rs:187]: I'm wondering if a zero-sized buffer is the right call. Theoretically that means we can't differentiate between an EOF `Ok(0)` and a "there was stuff to read but you gave me a zero-byte buffer to write into" `Ok(0)`. We don't actually know what to do differently in the second case, so maybe it doesn't matter?

Comment (LucioFranco) [reply src/sinks/util/tcp.rs:187]: I agreed, I changed this to a size 1 buffer. If we do somehow get some bytes on the read side we can just ignore it.

File: src/sinks/socket.rs (hunk 2 above)
Comment (LucioFranco) [reply src/sinks/socket.rs:238]: Yes, but I really liked the conceptual idea to shutdown the server outside of a count. Aka driven from outside. I'd prefer to keep it this way, if we find its bad in a refactor we can change it then.

File: src/sinks/socket.rs (hunk 3 above)
Comment (LucioFranco) [reply src/sinks/socket.rs:242]: We need to sleep to ensure the server read all 10 events. This kinda simulates an idle timeout. So we want to ensure the 10 got there before shutdown then shutdown then send another 10 which should trigger the reconnect process. Then check another 10.

File: src/sinks/socket.rs (hunk 2 above)
Comment (lukesteensen) [reply src/sinks/socket.rs:238]: I'm just curious why. In the real world, it is the server's decision when to shut down the connection and doing it here based on the count seems in line with that? And it also addresses the problem of sleeping to ensure the server had read all ten.

Comment (LucioFranco) [reply src/sinks/socket.rs:238]: If we don't have this notify/count how would the client know when to send the next 10 events? Does it sleep for a period before pumping new events?

File: src/sinks/util/tcp.rs [hunk 5]
```diff
@@ -175,14 +175,33 @@ impl Sink for TcpSink {
...
+                // If this returns `WouldBlock` we know the connection is still
+                // valid and the write will most likely succeed.
+                match conn.read(&mut [0u8; 1]) {
                     Err(error) if error.kind() != ErrorKind::WouldBlock => {
                         emit!(TcpConnectionDisconnected { error });
                         self.state = TcpSinkState::Disconnected;
                         Ok(AsyncSink::NotReady(line))
                     }
+                    Ok(0) => {
+                        emit!(TcpConnectionShutdown {});
```
Comment (binarylogic) [inline src/sinks/util/tcp.rs:201]: 🤩

File: src/sinks/socket.rs [hunk 6]
```diff
@@ -155,4 +155,106 @@ mod test {
...
+            }),
+        };
+        let mut rt = runtime();
+        let context = SinkContext::new_test(rt.executor());
+        let (sink, _healthcheck) = config.build(context).unwrap();
+
+        use std::sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc,
+        };
```
Comment (bruceg) [inline src/sinks/socket.rs:187]: Perhaps these should go with the other `use`s at the top of the module.

File: src/sinks/util/tcp.rs (hunk 4 above)
Comment (bruceg) [inline src/sinks/util/tcp.rs:187]: It's no longer a 0 sized buffer.

File: src/tls/maybe_tls.rs [hunk 7]
```diff
@@ -32,6 +32,13 @@ impl<R, T> MaybeTls<R, T> {
             Self::Tls(tls) => Some(&tls),
         }
     }
+
+    pub fn tls_mut(&mut self) -> Option<&mut T> {
+        match self {
+            Self::Raw(_) => None,
+            Self::Tls(tls) => Some(tls),
+        }
+    }
```
Comment (bruceg) [inline src/tls/maybe_tls.rs:41]: I don't see this used anywhere in this PR. Did this slip in by accident?

**Pull request merged**